lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,331 @@
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(data, seed, NimbleOptions.validate!(opts, @opts_schema))
+  end
+
+  defnp fit_n(data, seed, opts) do
+    data = to_float(data)
+    iterations = opts[:iterations]
+    damping_factor = opts[:damping_factor]
+    self_preference = opts[:self_preference]
+
+    {initial_a, initial_r, s, affinity_matrix} =
+      initialize_matrices(data, self_preference: self_preference)
+
+    {n, _} = Nx.shape(initial_a)
+    key = Nx.Random.key(seed)
+    {normal, _new_key} = Nx.Random.normal(key, 0, 1, shape: {n, n}, type: Nx.type(s))
+
+    s =
+      s +
+        normal *
+          (Nx.Constants.smallest_positive_normal(Nx.type(s)) * 100 +
+             Nx.Constants.epsilon(Nx.type(data)) / 10 * s)
+
+    range = Nx.iota({n})
+
+    {a, r, _, _, _} =
+      while {a = initial_a, r = initial_r, s, range, i = 0},
+            i < iterations do
+        temp = a + s
+        indices = Nx.argmax(temp, axis: 1)
+        y = Nx.reduce_max(temp, axes: [1])
+
+        neg_inf = Nx.Constants.neg_infinity(Nx.Type.to_floating(Nx.type(a)))
+        neg_infinities = Nx.broadcast(neg_inf, {n})
+        max_indices = Nx.stack([range, indices], axis: 1)
+        temp = Nx.indexed_put(temp, max_indices, neg_infinities)
+        y2 = Nx.reduce_max(temp, axes: [1])
+
+        temp = s - Nx.new_axis(y, -1)
+        temp = Nx.indexed_put(temp, max_indices, Nx.gather(s, max_indices) - y2)
+        temp = temp * (1 - damping_factor)
+        r = r * damping_factor + temp
+
+        temp = Nx.max(r, 0)
+        temp = Nx.put_diagonal(temp, Nx.take_diagonal(r))
+        temp = temp - Nx.sum(temp, axes: [0])
+        a_change = Nx.take_diagonal(temp)
+
+        temp = Nx.max(temp, 0)
+        temp = Nx.put_diagonal(temp, a_change)
+        temp = temp * (1 - damping_factor)
+        a = a * damping_factor - temp
+
+        {a, r, s, range, i + 1}
+      end
+
+    diagonals = Nx.take_diagonal(a) + Nx.take_diagonal(r) > 0
+
+    k = Nx.sum(diagonals, axes: [0])
+    {n, _} = shape = Nx.shape(data)
+
+    {cluster_centers, cluster_centers_indices, labels} =
+      if k > 0 do
+        mask = diagonals != 0
+
+        indices =
+          Nx.select(mask, Nx.iota(Nx.shape(diagonals)), -1)
+          |> Nx.as_type({:s, 64})
+
+        cluster_centers =
+          Nx.select(
+            Nx.broadcast(Nx.new_axis(mask, -1), shape),
+            data,
+            Nx.Constants.infinity(Nx.Type.to_floating(Nx.type(a)))
+          )
+
+        labels =
+          Nx.broadcast(mask, Nx.shape(s))
+          |> Nx.select(s, Nx.Constants.neg_infinity(Nx.type(s)))
+          |> Nx.argmax(axis: 1)
+          |> Nx.as_type({:s, 64})
+
+        {labels, _} =
+          while {labels, j = 0}, i <- indices do
+            if i >= 0 do
+              {Nx.put_slice(labels, [i], Nx.new_axis(j, -1)), j + 1}
+            else
+              {labels, j + 1}
+            end
+          end
```

Comment:
I have no idea what's going on, but indices are based off of the mask, so this should be equivalent:

```suggestion
        labels = Nx.select(mask, Nx.iota(Nx.shape(labels)), labels)
```